staging/src/k8s.io/client-go/rest/transport.go

```diff
@@ -88,6 +90,12 @@ func (c *Config) TransportConfig() (*transport.Config, error) {
 		Dial: c.Dial,
 	}
 
+	// If key and cert files are provided, but not keydata or certdata, then we should enable file reload logic by
+	// setting a callback here.
+	if len(c.KeyFile) > 0 && len(c.CertFile) > 0 && len(c.KeyData) == 0 && len(c.CertData) == 0 {
+		conf.TLS.GetCert = cachingCertificateLoader(c.CertFile, c.KeyFile)
```

Comment:
This is setting a dynamic GetCert for *any* client config that uses certificates. This is cache-busting for the client-go rest config tlsCache (function pointers aren't comparable, cert data or filenames are). That means every TransportFor() -> TransportConfig() invocation will return a config that results in a different underlying http.Transport when constructed (with its own pool of kept-alive connections).

As a point of reference, constructing the kubernetes clientset with NewForConfig() invokes ~40 of these, which means 40x the number of open sockets (and, more concerning, no bound on the roundtrippers we construct if we have poor client reuse hygiene)

See original discussion in https://github.com/kubernetes/kubernetes/pull/11820

Reply:
Is this cache busting? In my testing, I found that the function pointer is always the same irrelevant of input. We keep the cert filenames in place, so I think we'll end up with consistent cache keys? The cache keys will change when the certs rotate, though, because we use the cert data to form the key

Reply:
maybe its a risk relying on that - but the essence of what we're doing here requires cert files being converted to callbacks. perhaps we can come up with a way to be confident that the cache doesn't get busted, eg by removing the function pointer cache key element?

Reply:
if a dynamic cert function is provided, we *have* to include it in the cache key or we risk sharing transports between two configurations that actually don't use the same credentials.

one possibility is to move the certfile/keyfile/certdata/keydata -> dynamic loader transform into the place we construct the transport *after* determining the cache key for the config.

Reply:
thats a good idea - do you think we should worry about the cache key changing when file contents change, too?

Reply:
Hmm... need to think about that.

Today, we load referenced files and compute the cache key based on actual content, so certdata/certfile cases are normalized and handled the same way, and the cache key is based on the TLS data actually used by the client.

Today, [RESTClientFor](https://github.com/kubernetes/kubernetes/blob/release-1.17/staging/src/k8s.io/client-go/rest/config.go#L278-L298) does the following:
1. builds a `*transport.Config` by copying over the `*rest.Config` fields
2. populates transport.Config TLS data fields from transport.Config TLS file fields (we lose signal here whether the files should be preferred/used to refresh data)
3. computes cache key based on TLS data fields

That has the following implications:
* Step 1 ensures the original config remains untouched (which is good)
* Step 2 loses information about whether we should prefer file fields over data fields (which never mattered before because we didn't refresh file data, but now it does)
* Step 3 ignores the specified files and only looks at the loaded data (which never mattered before because we didn't refresh file data)

I think we need the following:
* A bit on the transport config that indicates we should reload from TLS files, even in the presence of TLS data fields, and set it to true when we load data from files into the config struct (maybe in TLSConfigFor or loadTLSFiles)
* In the cache key computation, if the ReloadTLSFiles bit is true, include the TLS filenames from the config in the cache key
* In TLSConfigFor, after computing the cache key, let the ReloadTLSFiles bit inform the `tlsConfig.GetClientCertificate` we set up

Reply:
> we lose signal here whether the files should be preferred/used to refresh data

that's what led to you needing to reverse/break config behavior in https://github.com/kubernetes/kubernetes/pull/79083/files#r360531199 ... I think the "ReloadTLSFiles" bit lets us maintain compatibility

Reply:
This sounds doable! I'm on it 🔬